// graphile-build-pg/node7minus/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.parseTags = exports.upperCamelCase = exports.constantCase = exports.camelCase = exports.upperFirst = exports.formatInsideUnderscores = exports.constantCaseAll = undefined;

var _toConsumableArray2 = require("babel-runtime/helpers/toConsumableArray");

var _toConsumableArray3 = _interopRequireDefault(_toConsumableArray2);

var _assign = require("babel-runtime/core-js/object/assign");

var _assign2 = _interopRequireDefault(_assign);

var _slicedToArray2 = require("babel-runtime/helpers/slicedToArray");

var _slicedToArray3 = _interopRequireDefault(_slicedToArray2);

var _upperFirst = require("lodash/upperFirst");

var _upperFirst2 = _interopRequireDefault(_upperFirst);

var _camelCase = require("lodash/camelCase");

var _camelCase2 = _interopRequireDefault(_camelCase);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var constantCaseAll = exports.constantCaseAll = function constantCaseAll(str) {
  return str.replace(/[^a-zA-Z0-9_]+/g, "_").replace(/[A-Z]+/g, "_$&").replace(/__+/g, "_").replace(/^[^a-zA-Z0-9]+/, "").replace(/^[0-9]/, "_$&") // GraphQL enums must not start with a number
  .toUpperCase();
};

var formatInsideUnderscores = exports.formatInsideUnderscores = function formatInsideUnderscores(fn) {
  return function (str) {
    var matches = str.match(/^(_*)([\s\S]*?)(_*)$/);
    if (!matches) {
      throw new Error("Impossible?"); // Satiate Flow
    }

    var _matches = (0, _slicedToArray3.default)(matches, 4),
        start = _matches[1],
        middle = _matches[2],
        end = _matches[3];

    return `${start}${fn(middle)}${end}`;
  };
};

var upperFirst = exports.upperFirst = formatInsideUnderscores(_upperFirst2.default);
var camelCase = exports.camelCase = formatInsideUnderscores(_camelCase2.default);
var constantCase = exports.constantCase = formatInsideUnderscores(constantCaseAll);
var upperCamelCase = exports.upperCamelCase = function upperCamelCase(str) {
  return upperFirst(camelCase(str));
};

var parseTags = exports.parseTags = function parseTags(str) {
  return str.split(`\n`).reduce(function (prev, curr) {
    if (prev.text !== "") {
      return (0, _assign2.default)({}, prev, {
        text: `${prev.text}\n${curr}`
      });
    }
    var match = curr.match(/^@[a-zA-Z][a-zA-Z0-9_]*($|\s)/);
    if (!match) {
      return (0, _assign2.default)({}, prev, {
        text: curr
      });
    }
    var key = match[0].substr(1).trim();
    var value = match[0] === curr ? true : curr.replace(match[0], "");
    return (0, _assign2.default)({}, prev, {
      tags: (0, _assign2.default)({}, prev.tags, {
        [key]: !prev.tags.hasOwnProperty(key) ? value : Array.isArray(prev.tags[key]) ? [].concat((0, _toConsumableArray3.default)(prev.tags[key]), [value]) : [prev.tags[key], value]
      })
    });
  }, {
    tags: {},
    text: ""
  });
};
