// graphile-build-pg/node7minus/plugins/PageInfoStartEndCursor.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

exports.default = function PageInfoStartEndCursor(builder) {
  builder.hook("GraphQLObjectType:fields", function (fields, _ref, _ref2) {
    var extend = _ref.extend,
        getTypeByName = _ref.getTypeByName;
    var Self = _ref2.Self,
        fieldWithHooks = _ref2.fieldWithHooks;

    if (Self.name !== "PageInfo") {
      return fields;
    }
    var Cursor = getTypeByName("Cursor");
    return extend(fields, {
      startCursor: fieldWithHooks("startCursor", function (_ref3) {
        var addDataGenerator = _ref3.addDataGenerator;

        addDataGenerator(function () {
          return { usesCursor: [true] };
        });
        return {
          description: "When paginating backwards, the cursor to continue.",
          type: Cursor
        };
      }, {
        isPageInfoStartCursorField: true
      }),
      endCursor: fieldWithHooks("endCursor", function (_ref4) {
        var addDataGenerator = _ref4.addDataGenerator;

        addDataGenerator(function () {
          return { usesCursor: [true] };
        });
        return {
          description: "When paginating forwards, the cursor to continue.",
          type: Cursor
        };
      }, {
        isPageInfoStartCursorField: true
      })
    }, `Adding startCursor/endCursor to ${Self.name}`);
  });
};
