// graphile-build-pg/node7minus/plugins/PgMutationProceduresPlugin.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _makeProcField = require("./makeProcField");

var _makeProcField2 = _interopRequireDefault(_makeProcField);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = function PgMutationProceduresPlugin(builder) {
  builder.hook("GraphQLObjectType:fields", function (fields, build, _ref) {
    var isRootMutation = _ref.scope.isRootMutation,
        fieldWithHooks = _ref.fieldWithHooks;

    if (!isRootMutation) {
      return fields;
    }
    var extend = build.extend,
        introspectionResultsByKind = build.pgIntrospectionResultsByKind,
        inflection = build.pgInflection;

    return extend(fields, introspectionResultsByKind.procedure.filter(function (proc) {
      return !proc.isStable;
    }).filter(function (proc) {
      return !!proc.namespace;
    }).reduce(function (memo, proc) {
      /*
      proc =
        { kind: 'procedure',
          name: 'integration_webhook_secret',
          description: null,
          namespaceId: '6484381',
          isStrict: false,
          returnsSet: false,
          isStable: true,
          returnTypeId: '2950',
          argTypeIds: [ '6484569' ],
          argNames: [ 'integration' ],
          argDefaultsNum: 0 }
      */

      var fieldName = inflection.functionName(proc.name, proc.namespace.name);
      memo[fieldName] = (0, _makeProcField2.default)(fieldName, proc, build, {
        fieldWithHooks,
        isMutation: true
      });
      return memo;
    }, {}), `Adding mutation procedure to root Mutation field`);
  });
};
